### PR TITLE
eth: Add stm32 ethernet promisc mode capabilities to hal

### PR DESCRIPTION
**Pull request by Rushybrook**
just adds the definitions required to the stm32 ethernet HAL add promisc mode support

### CODE REVIEW

**Review by erwango — CHANGES_REQUESTED** [review]
Thanks for this contribution.
Before going into the pure code review, please have a check to the [contributions guidelines](https://docs.zephyrproject.org/latest/contribute/index.html#commit-guidelines) and review the commit message.
Also, please refactor the branch in order to remove the merge commit.

File: drivers/ethernet/eth_stm32_hal.c [hunk 1]
```diff
@@ -846,6 +849,15 @@ static int eth_stm32_hal_set_config(const struct device *dev,
 				     sizeof(dev_data->mac_addr),
 				     NET_LINK_ETHERNET);
 		return 0;
+#if defined(CONFIG_NET_PROMISCUOUS_MODE)
+	case ETHERNET_CONFIG_TYPE_PROMISC_MODE:
+		dev_data = DEV_DATA(dev);
+		heth = &dev_data->heth;
```
Comment (erwango) [inline drivers/ethernet/eth_stm32_hal.c:855]: This could be factorized at the beginning of function.

**Review by erwango — CHANGES_REQUESTED** [review]
Thanks for this change.
Minor nitpick otherwise seems good

File: drivers/ethernet/eth_stm32_hal.c (hunk 1 above)
Comment (Rushybrook) [reply drivers/ethernet/eth_stm32_hal.c:855]: Agreed, I'll pull the assignments for `dev_data` and `heth` out of the switch statements for both cases and put them just under the declarations.

File: drivers/ethernet/eth_stm32_hal.c [hunk 2]
```diff
@@ -830,11 +833,11 @@ static int eth_stm32_hal_set_config(const struct device *dev,
 	struct eth_stm32_hal_dev_data *dev_data;
```
Comment (Nukersson) [inline drivers/ethernet/eth_stm32_hal.c:833]:
```suggestion
	int ret = -ENOTSUP;
	struct eth_stm32_hal_dev_data *dev_data;
```

You will need it later.

File: drivers/ethernet/eth_stm32_hal.c [hunk 3]
```diff
@@ -846,6 +849,12 @@ static int eth_stm32_hal_set_config(const struct device *dev,
 				     sizeof(dev_data->mac_addr),
 				     NET_LINK_ETHERNET);
 		return 0;
```
Comment (Nukersson) [inline drivers/ethernet/eth_stm32_hal.c:851]:
```suggestion
		ret = 0;
		break;
```

File: drivers/ethernet/eth_stm32_hal.c [hunk 4]
```diff
@@ -846,6 +849,12 @@ static int eth_stm32_hal_set_config(const struct device *dev,
...
+#if defined(CONFIG_NET_PROMISCUOUS_MODE)
+	case ETHERNET_CONFIG_TYPE_PROMISC_MODE:
+		heth->Instance->MACFFR |= config->promisc_mode ? ETH_MACFFR_PM : ~ETH_MACFFR_PM;
+
+		return 0;
+#endif
```
Comment (Nukersson) [inline drivers/ethernet/eth_stm32_hal.c:857]:
```suggestion
	case ETHERNET_CONFIG_TYPE_PROMISC_MODE:
#if defined(CONFIG_NET_PROMISCUOUS_MODE)
#if defined(CONFIG_SOC_SERIES_STM32H7X)
		heth->Instance->MACPFR |= config->promisc_mode ? ETH_MACPFR_PR : ~ETH_MACPFR_PR;
#else
		heth->Instance->MACFFR |= config->promisc_mode ? ETH_MACFFR_PM : ~ETH_MACFFR_PM;
#endif  /* CONFIG_SOC_SERIES_STM32H7X */
		ret = 0;
#endif /* CONFIG_NET_PROMISCUOUS_MODE */
		break;
```

Changes here so far:

1.  case should be always active
2. return -ENOSUP if `ETHERNET_CONFIG_TYPE_PROMISC_MODE` is given, but not enabled
3. add support for stm32h7 also

**Review by KozhinovAlexander — CHANGES_REQUESTED** [review]
Please add suggested changes to support stm32h7 also and rebase over current main repo. Please also provide proper testing in the future!
@erwango ping

File: drivers/ethernet/eth_stm32_hal.c (hunk 4 above)
Comment (Rushybrook) [reply drivers/ethernet/eth_stm32_hal.c]: Thanks!
This must have been overlooked in review and obviously never disabled by us (or just lucky)
happy with all your suggestions and thanks again

**Review by KozhinovAlexander — APPROVED** [review]
LGTM now

File: drivers/ethernet/eth_stm32_hal.c [hunk 5]
```diff
@@ -845,7 +849,26 @@ static int eth_stm32_hal_set_config(const struct device *dev,
...
+#else
+		heth->Instance->MACFFR |= ETH_MACFFR_PM;
+#endif  /* CONFIG_SOC_SERIES_STM32H7X */
+               } else {
+#if defined(CONFIG_SOC_SERIES_STM32H7X)
+		heth->Instance->MACPFR &= ~ETH_MACPFR_PR;
+#else
+		heth->Instance->MACFFR &= ~ETH_MACFFR_PM;
+#endif  /* CONFIG_SOC_SERIES_STM32H7X */
+               }
```
Comment (erwango) [inline drivers/ethernet/eth_stm32_hal.c:868]: The following would be more readable:

```suggestion
#if defined(CONFIG_NET_PROMISCUOUS_MODE)
#if defined(CONFIG_SOC_SERIES_STM32H7X)
               if(config->promisc_mode) {
               		heth->Instance->MACPFR |= ETH_MACPFR_PR;
               } else {
               		heth->Instance->MACPFR &= ~ETH_MACPFR_PR;
               }
#else
                if(config->promisc_mode) {
               		 heth->Instance->MACFFR |= ETH_MACFFR_PM;
               } else {
               		heth->Instance->MACFFR &= ~ETH_MACFFR_PM;
               }
#endif  /* CONFIG_SOC_SERIES_STM32H7X */
#endif /* CONFIG_NET_PROMISCUOUS_MODE */
```

Comment (Nukersson) [reply drivers/ethernet/eth_stm32_hal.c:868]: Agree with it. Thanks. Yesterday it was late night, as I did testings and changes.

Comment (erwango) [reply drivers/ethernet/eth_stm32_hal.c:868]: @Nukersson that's why we're multiple eyeballs in reviews. Thanks for your valuable review anyway.

**Comment by Nukersson** [comment]
@Rushybrook Please add changes suggested by @erwango and were are done here. Thank you for your contribution :)

**Review by erwango — APPROVED** [review]

**Review by jukkar — APPROVED** [review]

**Pull request merged**